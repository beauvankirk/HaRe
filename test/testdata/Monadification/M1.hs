module M1 where

f :: [a] -> b -> b
f [] y = y
f x y = f (g x) (h y)

g :: [a] -> [a]
g x = tail x

h :: a -> a
h x = id x 
