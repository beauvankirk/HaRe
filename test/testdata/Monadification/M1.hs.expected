module M1 where

f :: Monad m => [a] -> b -> m b
f [] y = return y
f x y = (h y) >>= (\hare1 -> f (g x) hare1)

g :: [a] -> [a]
g x = tail x

h :: Monad m => a -> m a
h x = return (id x)
